be/src/vec/columns/column_dictionary.h

```diff
@@ -0,0 +1,381 @@
+        }
+        return Status::OK();
+    }
+
+    void replace_column_data(const IColumn&, size_t row, size_t self_row = 0) override {
+        LOG(FATAL) << "should not call replace_column_data in ColumnDictionary";
+    }
+
+    void replace_column_data_default(size_t self_row = 0) override {
+        LOG(FATAL) << "should not call replace_column_data_default in ColumnDictionary";
+    }
+
+    void insert_many_dict_data(const int32_t* data_array, size_t start_index,
+                               const StringRef* dict_array, size_t data_num,
+                               uint32_t dict_num) override {
+        if (!is_dict_inited()) {
+            dict.reserve(dict_num);
+            for (uint32_t i = 0; i < dict_num; ++i) {
+                auto value = StringValue(dict_array[i].data, dict_array[i].size);
+                dict.insert_value(value);
+            }
+            dict_inited = true;
+        }
+
+        char* end_ptr = (char*)codes.get_end_ptr();
+        memcpy(end_ptr, data_array + start_index, data_num * sizeof(T));
+        end_ptr += data_num * sizeof(T);
+        codes.set_end_ptr(end_ptr);
+    }
+
+    bool is_dict_inited() const { return dict_inited; }
+
+    bool is_dict_sorted() const { return dict_sorted; }
+
+    bool is_dict_code_converted() const { return dict_code_converted; }
+
+    ColumnPtr convert_to_predicate_column() {
+        auto res = vectorized::PredicateColumnType<StringValue>::create();
+        size_t size = codes.size();
+        res->reserve(size);
+        for (size_t i = 0; i < size; ++i) {
+            auto& code = reinterpret_cast<T&>(codes[i]);
+            auto value = dict.get_value(code);
+            res->insert_data(value.ptr, value.len);
+        }
+        dict.clear();
+        return res;
+    }
+
+    void convert_dict_codes() {
+        if (!is_dict_sorted()) {
+            sort_dict();
+        }
+
+        if (!is_dict_code_converted()) {
+            for (size_t i = 0; i < size(); ++i) {
+                codes[i] = dict.convert_code(codes[i]);
+            }
+            dict_code_converted = true;
+        }
+    }
+
+    void sort_dict() {
+        dict.sort();
+        dict_sorted = true;
+    }
+
+    class Dictionary {
+    public:
+        Dictionary() = default;
+
+        void reserve(size_t n) {
+            dict_data.reserve(n);
+            inverted_index.reserve(n);
+        }
+
+        inline void insert_value(StringValue& value) {
+            dict_data.push_back_without_reserve(value);
+            inverted_index[value] = inverted_index.size();
+        }
+
+        inline T find_code(const StringValue& value) const {
+            auto it = inverted_index.find(value);
+            if (it != inverted_index.end()) {
+                return it->second;
+            }
+            return -1;
+        }
+
+        inline T find_bound_code(const StringValue& value, bool lower, bool eq) const {
+            if (lower) {
```

Comment:
Should we need to check whether value exists in dict_data?

Reply:
Check before function call. I refactored it to check inside the function.